developers/academy/js/starter_multimodal_data/102_mm_collections/30_import_data.mdx

```diff
@@ -0,0 +1,109 @@
+- Loops through the data and:
+    - Finds corresponding image to the text
+    - Converts the image to base64
+    - Adds objects to the batcher
+- Prints out any import errors
+
+## <i class="fa-solid fa-chalkboard"></i> Explain the code
+
+### <i class="fa-solid fa-chalkboard"></i> Preparation
+
+We use the native Node.js `fetch()` to load the data from the source, in this case a JSON file containing text data and a Zip file containing posters. The text data is then converted to a JSON object for easier manipulation and the images are extracted from the Zip file.
+
+Then, we create a collection object (with `client.collections.get`) so we can interact with the collection.
+
+### <i class="fa-solid fa-chalkboard"></i> Iterating over data
+
+The `for` loop is used in conjunction with `Object.keys()` to iterate through the elements in our JSON file. While iterating we increment the counter variable that lets us bulk insert objects in batches.
+
+<FilteredTextBlock
+  text={TSCode}
+  startMarker="// Iterate through data"
+  endMarker="// END Iterate through data"
+  language="ts"
+/>
+
+> We use `MAX_ITEMS` to limit the number of objects we import to our collection. Trail keys for the Cohere Embed 3 model [restrict](https://docs.cohere.com/v2/docs/rate-limits) you to 5 images per minute. 
+
+### <i class="fa-solid fa-chalkboard"></i> Add data to the Object
+
+#### Convert data types and build the Object
+
+The data is converted from a string to the correct data types for Weaviate. For example, the `release_date` is converted to a `Date` object, and the `genre_ids` are converted to a list of integers.
+
+<FilteredTextBlock
+  text={TSCode}
+  startMarker="// Format genre_ids and release_date"
+  endMarker="// Convert poster to base64"
+  language="ts"
+/>
+
+To save the image data as a `BLOB` (binary large object) data type, we convert the image to base64 using the helpful `toBase64FromMedia` utility that comes with the Weaviate client.
+
+<FilteredTextBlock
+  text={TSCode}
+  startMarker="// Convert poster to base64"
+  endMarker="// Insert"
+  language="ts"
+/>
+
+After converting data to the correct format, we build the object by its properties preparing it to be inserted into Weaviate.
+
+#### Bulk insert data
+
+Then we create on object that includes the uuid generated with `generateUuid5` from Weaviate and the object containing properties we previously define, we push this object to `itemsToInsert` for them to be bulk inserted with `insertMany()` once the batch is ready.
+
+<FilteredTextBlock
+  text={TSCode}
+  startMarker="// Insert"
+  endMarker="// END Insert"
+  language="ts"
+/>
+
+### <i class="fa-solid fa-chalkboard"></i> Error handling
+
+If you have any errors in your bulk insertion, you want to know something went wrong. That way you can decide how to handle them, such as by raising an exception. In this example, we simply print out that there was an error with the import.
```

Comment:
Is this what we want to do? It may be worth talking to someone like Tommy about this. 

It looks like if we insert 1000 objects and 1 fails this will throw an error. I'm not sure whether this is the kind of pattern we want users to use.

Reply:
yes, i've changed this to 20 objects but yes, we're working on getting an actual batcher to work around this.